src/diagonal.works/b6/expression.go

```diff
@@ -154,14 +151,15 @@ func (e Expression) ToProto() (*pb.NodeProto, error) {
 	return p, err
 }
 
-func (e *Expression) FromProto(node *pb.NodeProto) error {
+func ExpressionFromProto(node *pb.NodeProto) (Expression, error) {
+	e := Expression{}
 	switch n := node.Node.(type) {
 	case *pb.NodeProto_Symbol:
 		e.AnyExpression = new(SymbolExpression)
```

Comment:
Using `new` here is a bit confusing to me, since it returns a pointer (but then that pointer is replaced by a non-pointer when you call FromProto below).  Would it make sense to do something like:
```
var a AnyExpression
switch n := node.Node.(type) {
    case *pb.NodeProto_Symbol:
        a = SymbolExpression{}
    ....
}

if e, err = a.FromProto(node); err != nil {
    return Expression{}, err
}
......
```
Another alternative would be to remove FromProto from AnyExpression entirely, and replace it with a series of StringExpressionFromProto functions that are called from the type switch here (in the same way you changed FromProto on Expression to ExpressionFromProto, which I think was the correct thing to do).

Reply:
ye / done / sloppy / sorry